metaflow_extensions/deepspeed/plugins/deepspeed_libs/executor.py

```diff
@@ -0,0 +1,303 @@
+from metaflow.unbounded_foreach import UBF_CONTROL
+from metaflow.plugins.parallel_decorator import (
+    ParallelDecorator,
+    _local_multinode_control_task_step_func,
+)
+from metaflow.exception import MetaflowException
+import metaflow
+
+# from metaflow import Run as MetaflowRun
+from functools import partial
+from typing import List, Dict, Union, Tuple
+import subprocess
+import socket
+import json
+import time
+import sys
+import tempfile
+import os
+from io import BytesIO
+from collections import namedtuple
+from .exceptions import DeepspeedException
+from .datastore import DeepspeedDatastore
+from .status_notifier import (
+    TaskStatusNotifier,
+    HeartbeatThread,
+    wait_for_task_completion,
+    HeartbeatTimeoutException,
+    TaskFailedException,
+)
+
+DEEPSPEED_SUFFIX = "mf.deepspeed_datastore"
+
+
+HOSTFILE_IP_KEY = "hostfile_ips"
+HOSTFILE = "hostfile.txt"
+DEEPSPEED_JOB_COMPLETE_VAR = "mpi_job_status"
+PUBLIC_KEY_RECEIVED_VAR = "public_key_received"
+CONTROL_TASK_DONE_PATH = "control"
+MPI_PUBLIC_KEY_PATH = "public_keys"
+DEEPSPEED_ENV_FILE = ".deepspeed_env"  # https://github.com/microsoft/DeepSpeed/blob/24f20ef0a105d32f6085fe0d3b1c2f9324a6262c/docs/_tutorials/getting-started.md?plain=1#L230-L254
+DEEPSPEED_SUFFIX = "mf.deepspeed_datastore"
+
+
+def _get_path(local_output_dir, cloud_output_dir, node_index):
+    filepath_tuples = []
+    for path, subdirs, files in os.walk(local_output_dir):
+        for fname in files:
+            filepath_tuples.append(
+                (
+                    f"{cloud_output_dir}/{str(node_index)}/{os.path.relpath(os.path.join(path, fname), local_output_dir)}",
+                    os.path.join(path, fname),
+                )
+            )
+    return filepath_tuples
+
+
+class DeepspeedExecutor:
+    """
+    Instances of the DeepspeedExecutor class are responsible for launching the Deepspeed command. There is one per Metaflow @step annotated with @deepspeed.
+    DeepspeedExecutor consumes a list of hosts aka nodes aka k8s pods aka metaflow task containers, and information about how many processes to run on each.
```

Comment:
can be batch jobs too

Reply:
The executor is called within the user code. It represents `current.deepspeed`